src/notebook.c

```diff
@@ -754,6 +779,314 @@ void notebook_remove_page(gint page_num)
+	g_assert(child != NULL);
+	gtk_notebook_reorder_child(GTK_NOTEBOOK(main_widgets.notebook), child, pos);
+}
+
+
+void notebook_sort_tabs(NotebookTabSortMethod method)
+{
+	GArray *docs;
+	GeanyDocument* doc;
+	guint i;
+
+	docs = g_array_new(FALSE, TRUE, sizeof(GeanyDocument*));
+
+	foreach_document(i)
+		g_array_append_val(docs, documents[i]);
+
+	if (method == NOTEBOOK_TAB_SORT_FILENAME)
+		g_array_sort(docs, compare_docs_based_on_filename);
+	else
+		g_array_sort(docs, compare_docs_based_on_pathname);
+
+	for (i = 0; i < docs->len; ++i)
+	{
+		doc = g_array_index(docs, GeanyDocument*, i);
+		move_tab(doc, i);
+	}
+
+	g_array_free(docs, TRUE);
+}
+
+
+void on_sort_tabs_filename_activate(GtkMenuItem *menuitem, gpointer user_data)
+{
+	notebook_sort_tabs(NOTEBOOK_TAB_SORT_FILENAME);
+}
+
+
+void on_sort_tabs_pathname_activate(GtkMenuItem *menuitem, gpointer user_data)
+{
+	notebook_sort_tabs(NOTEBOOK_TAB_SORT_PATHNAME);
+}
+
+
+static void gradually_sort_tab_based_on_filename(GeanyDocument *doc)
+{
+	GeanyDocument *doc_b;
+	gchar *base_a, *base_b;
+	gint pos, i, n_pages;
+	GtkWidget *page;
+
+	base_a = get_doc_basename(doc);
+	pos = 0;
+	n_pages = gtk_notebook_get_n_pages(GTK_NOTEBOOK(main_widgets.notebook));
+
+	for (i = 0; i < n_pages; ++i)
+	{
+		page = gtk_notebook_get_nth_page(GTK_NOTEBOOK(main_widgets.notebook), i);
+		doc_b = document_get_from_notebook_child(page);
+
+		if (doc_b == doc)
+			continue;
+
+		base_b = get_doc_basename(doc_b);
+
+		if (compare_filenames(base_a, base_b) < 0)
+		{
+			move_tab(doc, pos);
+			g_free(base_b);
+			g_free(base_a);
+			return;
+		}
+
+		g_free(base_b);
+		++pos;
+	}
+
+	g_free(base_a);
+	move_tab(doc, -1);
+}
+
+
+static void gradually_sort_tab_based_on_pathname(GeanyDocument *doc)
+{
+	GeanyDocument *doc_b;
+	GtkWidget *page;
+	gchar *base_a, *base_b, *dirname_a, *dirname_b, *path_a, *path_b;
+	gint pos, i, n_pages, cmp;
+
+	dirname_a = get_doc_dirname(doc);
+	base_a = get_doc_basename(doc);
+	pos = 0;
+	n_pages = gtk_notebook_get_n_pages(GTK_NOTEBOOK(main_widgets.notebook));
+
+	for (i = 0; i < n_pages; ++i)
+	{
+		page = gtk_notebook_get_nth_page(GTK_NOTEBOOK(main_widgets.notebook), i);
+		doc_b = document_get_from_notebook_child(page);
+
+		if (doc_b == doc)
+			continue;
+
+		dirname_b = get_doc_dirname(doc_b);
+		cmp = compare_filenames(dirname_a, dirname_b);
+
+		if (cmp < 0)
+		{
+			g_free(dirname_b);
+			g_free(dirname_a);
+			g_free(base_a);
+			move_tab(doc, pos);
+			return;
+		}
+		else if (cmp == 0)
+		{
+			base_b = get_doc_basename(doc_b);
+
+			if (compare_filenames(base_a, base_b) < 0)
+			{
+				g_free(base_b);
+				g_free(dirname_b);
+				g_free(base_a);
+				g_free(dirname_a);
+				move_tab(doc, pos);
+				return;
+			}
+
+			g_free(base_b);
+		}
+
+		g_free(dirname_b);
+		++pos;
+	}
+
+	g_free(base_a);
+	g_free(dirname_a);
+	move_tab(doc, -1);
+}
+
+
+static void on_document_open(GObject *obj, GeanyDocument *doc)
+{
+	if (interface_prefs.show_notebook_tabs)
```

Comment:
this most likely is a bad case of premature optimization.  Because it the user had the tabs hidden but sorting enabled, and shows the tabs, it won't be sorted as he probably would have expected.